servlet/src/main/java/io/undertow/servlet/spec/ServletInputStreamImpl.java

```diff
@@ -196,50 +195,83 @@ public int read(final byte[] b, final int off, final int len) throws IOException
+                // this is unlikely to happen, it will happen only during timeouts and server shutdowns
+                if (anyAreSet(state, FLAG_FINISHED)) {
+                    return;
+                }
+                throw e;
+            }
+            int res = Channels.readBlocking(channel, byteBuffer);
+            byteBuffer.flip();
             if (res == -1) {
                 setFlags(FLAG_FINISHED);
-                pooled.close();
-                pooled = null;
+                closePoolIfNotNull();
             }
         }
     }
 
     private void readIntoBufferNonBlocking() throws IOException {
         if (pooled == null && !anyAreSet(state, FLAG_FINISHED)) {
-            pooled = bufferPool.allocate();
+            final ByteBuffer byteBuffer;
+            try {
+                PooledByteBuffer pooled = this.pooled = bufferPool.allocate();
+                byteBuffer = pooled.getBuffer();
+            } catch (NullPointerException e) {
+                // check for NPE + FLAG_FINISHED, it indicates a race condition where the buffer was closed and set to null by
+                // another thread
+                // instead of paying the price of synchronization, we just ignore the NPE and return, mimicking the code path
+                // we would follow in case the check in the if statement above returned false
+                // this is unlikely to happen, it will happen only during timeouts and server shutdowns
+                if (anyAreSet(state, FLAG_FINISHED)) {
+                    return;
+                }
+                throw e;
+            }
             if (listener == null) {
-                int res = channel.read(pooled.getBuffer());
-                if (res == 0) {
-                    pooled.close();
-                    pooled = null;
+                int res = channel.read(byteBuffer);
+                if (res == 0 && pooled != null) {
+                    closePoolIfNotNull();
                     return;
                 }
                 pooled.getBuffer().flip();
                 if (res == -1) {
                     setFlags(FLAG_FINISHED);
-                    pooled.close();
-                    pooled = null;
+                    closePoolIfNotNull();
                 }
             } else {
-                int res = channel.read(pooled.getBuffer());
+                int res = channel.read(byteBuffer);
                 pooled.getBuffer().flip();
                 if (res == -1) {
                     setFlags(FLAG_FINISHED);
-                    pooled.close();
-                    pooled = null;
+                    closePoolIfNotNull();
                 } else if (res == 0) {
                     clearFlags(FLAG_READY);
-                    pooled.close();
-                    pooled = null;
+                    closePoolIfNotNull();
                 }
             }
         }
     }
 
+    private void closePoolIfNotNull() {
+        try {
+            if (pooled != null) {
```

Comment:
ONly thing Id consider is to have local var? unless thats calculated choice to sometimes pay price of exception, rather than constant cost of local var and assignment.

Reply:
@fl4via  ^^

Reply:
We will pay the price of the NPE anyway, but, yes, I think adding the local var is a good idea because it will make the exception less likely.
`if (pooled != null) {
   final PooledByteBuffer pooled = this.pooled; // we can still read a null value here, hence, NPE when closing
   this.pooled = null;
   pooled.close();  // marginal risk of NPE here
} `

Reply:
As a secod thought, will it make it less likely, if the time spent seting it to a variable might be the same time it takes to invoke close directly? Or is it not? I'll try to see if I can find some information on this. But the code I suggested above might have the same statistical risk of getting the NPE as the one we have without the local variable.